### ISSUE TITLE
conversation with extension_id 65000 is rejected

### ISSUE THREAD

**Issue by NoName115** [openssl/openssl#7747]
When a ClientHello msg contain extension with id 65000 (cryptopro_bug extension), OpenSSL server does not parse the extension but reject this connection with `illegal_parameter` alert, as a bug in the client hello.
OpenSSL version: branch OpenSSL_1_1_1 - 1708e3e85b4a86bae26860aa5d2913fc8eff6086 , static build on Fedora27
Tested with tlsfuzzer:
```
# Run OpenSSL server
$ openssl s_server -bugs -www -key ../localhost.key -cert ../localhost.crt

# In another terminal
git clone https://github.com/tomato42/tlsfuzzer.git
pushd tlsfuzzer
git clone https://github.com/tomato42/tlslite-ng.git .tlslite-ng
pushd .tlslite-ng
popd
ln -s .tlslite-ng/tlslite tlslite
popd

# Run test-script
PYTHONPATH=tlsfuzzer scripts/test-tls13-large-number-of-extensions.py --supgroup
```

Expected output by tlsfuzzer:
```
Test end
  successful: 82
  failed: 0
```
Actual output:
```
# tlsfuzzer
Test end
successful: 80
failed: 2
  'empty unassigned extensions, ids in range from 61463 to 65535'
  'unassigned extensions with random payload, ids in range from 64535 to 65535'

# OpenSSL
Using default temp DH parameters
ACCEPT
140284767766336:error:141B306E:SSL routines:tls_collect_extensions:bad extension:ssl/statem/extensions.c:601:
140284767766336:error:141B306E:SSL routines:tls_collect_extensions:bad extension:ssl/statem/extensions.c:601:
```

**Comment by mattcaswell** [comment]
Arguably it's doing the right thing (sort of). That particular extension is an "unsolicited" extension that should only appear in the ServerHello and never in the ClientHello. So libssl is complaining because it has seen it appear in a message where it is not supposed to be.

OTOH, that extension was never officially registered and according to IANA is "unassigned":

https://www.iana.org/assignments/tls-extensiontype-values/tls-extensiontype-values.xhtml#tls-extensiontype-values-1

I've searched around and can't find any documentation for this extension to describe its behaviour (perhaps @beldmit might know because this was added for GOST support in commit 0e1dba93). Given the lack of a spec, and the unassigned status at IANA I am tempted to just change this to ignore the extension if it appears in a ClientHello - which is actually what the behaviour used to be before we tightened up the extensions logic.

Probably something like this would fix it (untested):
````
diff --git a/ssl/statem/extensions.c b/ssl/statem/extensions.c
index 63e61c6184..73e843f432 100644
--- a/ssl/statem/extensions.c
+++ b/ssl/statem/extensions.c
@@ -348,10 +348,12 @@ static const EXTENSION_DEFINITION ext_defs[] = {
     {
         /*
          * Special unsolicited ServerHello extension only used when
-         * SSL_OP_CRYPTOPRO_TLSEXT_BUG is set
+         * SSL_OP_CRYPTOPRO_TLSEXT_BUG is set. We allow it in a ClientHello but
+         * ignore it.
          */
         TLSEXT_TYPE_cryptopro_bug,
-        SSL_EXT_TLS1_2_SERVER_HELLO | SSL_EXT_TLS1_2_AND_BELOW_ONLY,
+        SSL_EXT_CLIENT_HELLO | SSL_EXT_TLS1_2_SERVER_HELLO
+        | SSL_EXT_TLS1_2_AND_BELOW_ONLY,
         NULL, NULL, NULL, tls_construct_stoc_cryptopro_bug, NULL, NULL
     },
     {
````

**Comment by beldmit** [comment]
I'm not sure this extension has any sense nowadays, with modern versions of Russian GOST CSPs.

**Comment by beldmit** [comment]
Well, the extension is still in use. The patch you provided does not solve the problem, I'll provide the patch later.

**Comment by beldmit** [comment]
Smoke test: 
```
OPENSSL_CONF=gost_engine.conf openssl s_client -connect tlsgost-2001.cryptopro.ru:443 -cipher GOST2001-GOST89-GOST89
```

**Comment by tomato42** [comment]
verified as fixed with 9effc496ad8a9
